engine/pkg/externalresource/s3/index.go

```diff
@@ -0,0 +1,309 @@
+		ret.backgroundTask()
+	}()
+	return ret
+}
+
+func (m *indexManagerImpl) SetPersisted(ctx context.Context, ident internal.ResourceIdent) (bool, error) {
+	if m.isClosed.Load() {
+		return false, errors.New("indexManager already closed")
+	}
+
+	if m.executorID != ident.Executor {
+		log.Panic("Cannot persist resource that is not created on the local executor",
+			zap.Any("ident", ident))
+	}
+
+	m.indexMu.Lock()
+	index, ok := m.indexFiles[ident.WorkerID]
+	if !ok {
+		index = newIndexFile()
+		m.indexFiles[ident.WorkerID] = index
+	}
+
+	ok = index.SetPersisted(ident)
+	m.indexMu.Unlock()
+	if !ok {
+		return false, nil
+	}
+
+	if err := m.triggerFlush(ctx, ident.WorkerID); err != nil {
+		return false, err
+	}
+	return true, nil
+}
+
+func (m *indexManagerImpl) LoadPersistedFileSet(
+	ctx context.Context, scope internal.ResourceScope,
+) (map[model.ResourceName]struct{}, error) {
+	storage, err := m.createStorageForIndexFile(ctx, scope)
+	if err != nil {
+		return nil, err
+	}
+
+	exists, err := storage.FileExists(ctx, indexFileName)
+	if err != nil {
+		return nil, errors.Annotate(err, "check index file exists")
+	}
+
+	if !exists {
+		return nil, errors.Trace(errIndexFileDoesNotExist)
+	}
+
+	bytes, err := storage.ReadFile(ctx, indexFileName)
+	if err != nil {
+		return nil, err
+	}
+	index := newIndexFile()
+	if err := index.DeserializeFrom(bytes); err != nil {
+		return nil, err
+	}
+
+	return index.GetPersistedFileSet(), nil
+}
+
+func (m *indexManagerImpl) Close() {
+	if m.isClosed.Swap(true) {
+		// Already closed
+		return
+	}
+	close(m.closeCh)
+	m.wg.Wait()
+}
+
+func (m *indexManagerImpl) backgroundTask() {
+	for {
+		select {
+		case <-m.closeCh:
+			return
+		case flushTask := <-m.flushIndexNotifyCh:
+			ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+			err := m.doFlush(ctx, flushTask.project, flushTask.workerID)
+			cancel()
+			flushTask.doneCh <- err
+		}
+	}
+}
+
+func (m *indexManagerImpl) triggerFlush(ctx context.Context, workerID model.WorkerID) error {
+	done := make(chan error, 1)
+	flushTask := struct {
+		project  tenant.ProjectInfo
+		workerID model.WorkerID
+		doneCh   chan error
+	}{
+		workerID: workerID,
+		doneCh:   done,
+	}
+
+	select {
+	case <-ctx.Done():
+		return errors.Trace(ctx.Err())
+	case m.flushIndexNotifyCh <- flushTask:
+	}
+
+	select {
+	case <-ctx.Done():
+		return errors.Trace(ctx.Err())
+	case err := <-done:
+		if err != nil {
+			return err
+		}
+		return nil
+	}
+}
+
+func (m *indexManagerImpl) doFlush(
+	ctx context.Context,
+	project tenant.ProjectInfo,
+	workerID model.WorkerID,
+) error {
+	m.indexMu.Lock()
+	index, ok := m.indexFiles[workerID]
+	if !ok {
+		m.indexMu.Unlock()
+		log.Info("indexManager: worker has been removed",
+			zap.String("worker-id", workerID))
+		return nil
+	}
+
+	bytes, err := index.Serialize()
+	m.indexMu.Unlock()
+	if err != nil {
+		return err
+	}
+
+	storage, err := m.createStorageForIndexFile(ctx, internal.ResourceScope{
```

Comment:
Is it necessary to create a new storage each time, could the index manager reuse one storage instance?